graphite-api/src/render.rs

```diff
@@ -0,0 +1,501 @@
+    } else {
+        // Absolute time
+        match s.as_str() {
+            "now" => Ok(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as u32),
+            "yesterday" => {
+                Ok(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as u32 - 3600 * 24)
+            }
+            "" => Err(ParseError::EmptyString),
+            // Unix timestamp parse as default
+            _ => {
+                // Unix timestamp
+                Ok(s.parse::<u32>()?)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum RenderFormat {
+    Png,
+    Raw,
+    Csv,
+    Json,
+    Svg,
+    Pdf,
+    Dygraph,
+    Rickshaw,
+}
+
+impl FromStr for RenderFormat {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "png" => Ok(RenderFormat::Png),
+            "raw" => Ok(RenderFormat::Raw),
+            "csv" => Ok(RenderFormat::Csv),
+            "json" => Ok(RenderFormat::Json),
+            "svg" => Ok(RenderFormat::Svg),
+            "pdf" => Ok(RenderFormat::Pdf),
+            "dygraph" => Ok(RenderFormat::Dygraph),
+            "rickshaw" => Ok(RenderFormat::Rickshaw),
+            _ => Err(ParseError::RenderFormat),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderMetric(String);
+
+#[derive(Debug, PartialEq)]
+pub struct RenderPath(PathBuf);
+
+fn path(dir: &Path, metric: &str) -> Result<PathBuf, Error> {
+    let path = metric
+        .split('.')
+        .fold(PathBuf::new(), |acc, x| acc.join(x))
+        .with_extension("wsp");
+    let full_path = dir.canonicalize()?.join(&path);
+    Ok(full_path)
+}
+
+fn walk(dir: &Path, metric: &str, q: &RenderQuery) -> Result<Vec<RenderPoint>, Error> {
+    let full_path = path(dir, metric)?;
+    let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as u32;
+    let interval = Interval::new(q.from, q.until).map_err(err_msg)?;
+
+    let archive = WhisperFile::open(&full_path)?.fetch_auto_points(interval, now)?;
+
+    let mut points = Vec::new();
+
+    for (index, value) in archive.values.iter().enumerate() {
+        let time = archive.from_interval + archive.step * index as u32;
+        points.push(RenderPoint(*value, time));
+    }
+
+    Ok(points)
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderPoint(Option<f64>, u32);
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderResponceEntry {
+    target: String,
+    datapoints: Vec<RenderPoint>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RenderResponce {
+    entries: Vec<Option<RenderResponceEntry>>,
+}
+
+pub fn render_handler(state: State<Args>, params: RenderQuery) -> Result<HttpResponse, Error> {
+    let dir = &state.path;
+    let response: Vec<RenderResponceEntry> = params
+        .target
+        .iter()
```

Comment:
into_iter()

Reply:
clippy considered that into_iter() should be replaced with .iter(), so I did it

Reply:
iter instead of into_iter is a cause of extra allocation in the loop (`target: x.to_string()`)

Reply:
and we received:
> borrow of moved value: `params`

Reply:
```
    let response: Vec<RenderResponceEntry> = params
        .target
        .clone()
        .into_iter()
        .map(|x| RenderResponceEntry {
            datapoints: walk(&dir, &x, &params).unwrap(),
            target: x,
        })
        .collect();
```
Is it better?

Reply:
remade it